Tools/scripts/summarize_stats.py

```diff
@@ -2,908 +2,988 @@
 
-def _emit_comparative_execution_counts(base_rows, head_rows):
-    base_data = {x[0]: x[1:] for x in base_rows}
-    head_data = {x[0]: x[1:] for x in head_rows}
-    opcodes = base_data.keys() | head_data.keys()
+        if len(data_a) != len(rows_a) or len(data_b) != len(rows_b):
+            raise ValueError("Duplicate keys")
 
-    rows = []
-    default = [0, "0.0%", "0.0%", 0]
-    for opcode in opcodes:
-        base_entry = base_data.get(opcode, default)
-        head_entry = head_data.get(opcode, default)
-        if base_entry[0] == 0:
-            change = 1
+        # To preserve ordering, use A's keys as is and then add any in B that
+        # aren't in A
+        keys = list(data_a.keys()) + [k for k in data_b.keys() if k not in data_a]
+        rows = [
+            self.join_row(k, data_a.get(k, default), data_b.get(k, default))
+            for k in keys
+        ]
+        if self.sort_by_last_column:
+            rows.sort(key=lambda row: abs(float(row[-1])), reverse=True)
+
+        columns = self.join_columns(self.columns)
+        return columns, rows
+
+    def get_rows(
+        self, base_stats: Stats, head_stats: Stats | None = None
+    ) -> tuple[Columns, Rows]:
+        if head_stats is None:
+            rows = self.calculate_rows(base_stats)
+            return self.columns, rows
         else:
-            change = (head_entry[0] - base_entry[0]) / base_entry[0]
-        rows.append((opcode, base_entry[0], head_entry[0], f"{change:0.1%}"))
+            rows_a = self.calculate_rows(base_stats)
+            rows_b = self.calculate_rows(head_stats)
+            cols, rows = self.join_tables(rows_a, rows_b)
+            return cols, rows
+
+    def output_markdown(
+        self,
+        out: TextIO,
+        base_stats: Stats,
+        head_stats: Stats | None = None,
+        level: int = 2,
+    ) -> None:
+        header, rows = self.get_rows(base_stats, head_stats)
+        if len(rows) == 0:
+            return
+
+        def to_markdown(x):
+            if hasattr(x, "markdown"):
+                return x.markdown()
+            elif isinstance(x, str):
+                return x
+            elif x is None:
+                return ""
+            else:
+                raise TypeError(f"Can't convert {x} to markdown")
+
+        width = len(header)
+        header_line = "|"
+        under_line = "|"
+        for item in header:
+            under = "---"
+            if item.endswith(":"):
+                item = item[:-1]
+                under += ":"
+            header_line += item + " | "
+            under_line += under + "|"
+        print(header_line, file=out)
+        print(under_line, file=out)
+        for row in rows:
+            if width is not None and len(row) != width:
+                raise ValueError("Wrong number of elements in row '" + str(row) + "'")
+            print("|", " | ".join(to_markdown(i) for i in row), "|", file=out)
+        print(file=out)
+
+
+class FixedTable(Table):
```

Comment:
I don't like all these subclasses of `Table`, it mixes up the responsibilities of creating the contents and the formatting.
Having a single `Table` class which does the formatting and takes the contents as a parameter to its `__init__` would separate the responsibilities better. 
So instead of `ExecutionCountTable("uops")`, it would be `Table("uops", get_uop_counts())`

Reply:
The data can support merging, etc. `get_uop_counts()` can return an object that supports that functionality.
I just think that separating format from data manipulation will make things more maintainable.